### ISSUE TITLE
Unable to open empty ROOT file with nested directories

### ISSUE THREAD

**Issue by reikdas** [scikit-hep/uproot#104]
When I try to read an empty ROOT file with nested directories, uproot throws an error on calling the .allkeys() method. 

```
>>> import uproot
>>> file = uproot.open("nested.root")
>>> file.allkeys()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/reikdas/uproot/uproot/rootio.py", line 311, in allkeys
    return self.keys(recursive=True, filtername=filtername, filterclass=filterclass)
  File "/home/reikdas/uproot/uproot/rootio.py", line 299, in keys
    return list(self.iterkeys(recursive=recursive, filtername=filtername, filterclass=filterclass))
  File "/home/reikdas/uproot/uproot/rootio.py", line 265, in iterkeys
    for name in key.get().iterkeys(recursive, filtername, filterclass):
  File "/home/reikdas/uproot/uproot/rootio.py", line 265, in iterkeys
    for name in key.get().iterkeys(recursive, filtername, filterclass):
  File "/home/reikdas/uproot/uproot/rootio.py", line 848, in get
    return ROOTDirectory.read(self._source, self._cursor.copied(), self._context, self)
  File "/home/reikdas/uproot/uproot/rootio.py", line 192, in read
    headerkey = TKey.read(source, subcursor, context, None)
  File "/home/reikdas/uproot/uproot/rootio.py", line 788, in read
    out = cls._readinto(out, source, cursor, context, parent)
  File "/home/reikdas/uproot/uproot/rootio.py", line 821, in _readinto
    raise ValueError("TKey declares that object {0} has {1} bytes but only {2} remain in the file".format(repr(self.fName), self.fNbytes, source.size() - self.fSeekKey))
ValueError: TKey declares that object b'\x00' has 1919905652 bytes but only -45677840 remain in the file
```
When trying to open the same via ROOT - 

```
[reikdas@lpcdt076 uproot]$ root -l nested.root
root [0] 
Attaching file nested.root as _file0...
(TFile *) 0xdaa940
root [1] _file0->ls();
TFile**		nested.root	
 TFile*		nested.root	
  KEY: TDirectoryFile	tof;1	tof
root [2] _file0->GetDirectory("tof")
(TDirectory *) 0x1601620
root [3] first = _file0->GetDirectory("tof");
root [4] _file0->GetDirectory("tof")->GetDirectory("in")
(TDirectory *) 0x1653140
root [5] second = _file0->GetDirectory("tof")->GetDirectory("in");
```

I have attached the C file used to create the ROOT file and the ROOT file itself. 
[nested.zip](https://github.com/scikit-hep/uproot/files/2202352/nested.zip)

Although, I do think no one is ever going to come across this as a use case.

**Comment by jpivarski** [comment]
True, but this sorry of thing is important to fix. We can fail to support certain seldom-used features, but we can't throw exceptions like this: especially since this was reading off the end of a buffer, which can yeild wrong results, rather than an exception.

`allkeys` calls the more fundamental `keys`, and it's probably something simple. (Not to mention relevant to what you're working on.)

**Comment by jpivarski** [comment]
The following is sufficient to trigger the error:

```python
>>> import uproot
>>> f = uproot.open("/tmp/downloads/nested.root")
>>> f["tof"]["in"]
```

**Comment by jpivarski** [comment]
The trouble is that we assume everything with class name "TDirectory" or "TDirectoryFile" is a `ROOTDirectory`, which means

```python
                fVersion, fDatimeC, fDatimeM, fNbytesKeys, fNbytesName = cursor.fields(source, ROOTDirectory._format3)
                if fVersion <= 1000:
                    fSeekDir, fSeekParent, fSeekKeys = cursor.fields(source, ROOTDirectory._format4_small)
                else:
                    fSeekDir, fSeekParent, fSeekKeys = cursor.fields(source, ROOTDirectory._format4_big)
```

This empty subdirectory has a different structure: it has a `TKey` structure

```python
        self.fNbytes, self.fVersion, self.fObjlen, self.fDatime, self.fKeylen, self.fCycle, self.fSeekKey, self.fSeekPdir = cursor.fields(source, self._format_small)
        if self.fVersion > 1000:
            cursor.index = start
            self.fNbytes, self.fVersion, self.fObjlen, self.fDatime, self.fKeylen, self.fCycle, self.fSeekKey, self.fSeekPdir = cursor.fields(source, self._format_big)

        self.fClassName = cursor.string(source)
        self.fName = cursor.string(source)
        self.fTitle = cursor.string(source)
```

followed by a `ROOTDirectory` structure with `fNbytesKeys == 0` and `fSeekKeys == 0`. It would be easy to put in a special case to interpret `fNbytesKeys == 0` or `fSeekKeys == 0` as an empty directory, but somehow you have to know that the "in" directory has an extra `TKey` header before its `ROOTDirectory`. This business of having an extra `TKey` header may be unrelated to the fact that it's empty, and could therefore be an important case for non-empty directories that we somehow haven't encountered. It's entirely possible that it happens for all directories with deeper nesting than two.

**Comment by jpivarski** [comment]
Actually, it turned out to be simpler than I indicated above; my debugging was confused (so I reverted to the GitHub version and tried again). All we need is a check for `fSeekKeys == 0` to identify an empty directory; we shouldn't really seek to the beginning of the file to look for its keys. With this one correction, we have no trouble reading empty directories.

So to cancel the false alarm: there doesn't seem to be two different directory types requiring different byte-handling. I only see one directory type. Confused debugging was the culprit (as it usually is).
